ai2thor/controller.py

```diff
@@ -952,25 +952,84 @@ def _start_unity_thread(self, env, width, height, server_params, image_name):
+                disp = Xlib.display.Display(":%s" % display_str)
+                for screen in range(0, disp.screen_count()):
+                    disp_screen_str = ":%s.%s" % (display_str, screen)
+
+                    disp_screen = Xlib.display.Display(disp_screen_str)
+
+                    if "GLX" in disp_screen.list_extensions():
+                        valid_displays.append(disp_screen_str)
+
+            except Xlib.error.DisplayConnectionError as e:
+                logger.warning(
+                    "could not connect to X Display: %s, %s" % (display_str, e)
+                )
+
+        return valid_displays
+
+    def _select_x_display(self):
+
+        valid_displays = self._valid_x_displays()
+        if valid_displays:
+            return random.choice(valid_displays)
+        else:
+            # return 0.0 as default, an error will be thrown from check_display fails
+            return ":0.0"
+
+    def _valid_displays_message(self):
+        valid_displays = self._valid_x_displays()
+        if valid_displays:
+            return "The following valid displays were found %s" % (
+                ", ".join(valid_displays)
+            )
+        else:
+            return ""
+
     def check_x_display(self, x_display):
-        with open(os.devnull, "w") as dn:
-            # copying the environment so that we pickup
-            # XAUTHORITY values
-            env = os.environ.copy()
-            env["DISPLAY"] = x_display
-
-            if subprocess.call(["which", "xdpyinfo"], stdout=dn) == 0:
-                assert (
-                    subprocess.call("xdpyinfo", stdout=dn, env=env, shell=True) == 0
-                ), (
-                    "Invalid DISPLAY %s - cannot find X server with xdpyinfo"
-                    % x_display
+        import Xlib.display
+
+        try:
+            display = Xlib.display.Display(x_display)
+            screen_parts = x_display.split(".")
+            if len(screen_parts) > 1:
+                # this Xlib.display will find a valid screen if an
+                # invalid screen was passed in (e.g. :0.9999999 -> :0.1)
+                if screen_parts[1] != str(display.get_default_screen()):
+                    raise ValueError(
+                        "Invalid display, non-existent screen: %s; %s"
+                        % (x_display, self._valid_displays_message())
+                    )
+
+            if "GLX" not in display.list_extensions():
+                valid_displays = self._valid_x_displays()
+                raise ValueError(
+                    "Display %s does not have the GLX extension loaded.  GLX is required by Unity3D. %s"
+                    % (x_display, self._valid_displays_message())
                 )
 
+        except (Xlib.error.DisplayNameError, Xlib.error.DisplayConnectionError) as e:
+            raise ValueError(
+                "Invalid display: %s - %s. %s"
+                % (x_display, e, self._valid_displays_message())
+            )
```

Comment:
Can we add a little more information to this error message? In particular I think it would be great if it told people what the problem was (i.e. either "we could not find a valid x_display" or "the given x_display was bad") and gave them some information about how to trouble shoot / fix this (i.e. "If no X-display is currently running on your machine, please run our `startx.py` (link) script as `sudo python startx.py &` which will start X-servers on each of your GPUs. Note that you will need administrator privileges to start these X-servers, if you do not have such privileges please contact your server administrator.".